src/main/scala/au/com/agiledigital/toolform/reader/ProjectReader.scala

```diff
@@ -34,13 +42,44 @@ object ProjectReader {
           val config = simpleConfig.resolve()
 
           val projectResult = loadConfig[Project](config)
-          resultOrCollectReadErrors(projectResult)
+
+          for {
+            collectedReadErrors    <- resultOrCollectReadErrors(projectResult)
+            validatedProjectResult <- validateProjectResult(collectedReadErrors)
+          } yield validatedProjectResult
+
         case Failure(e) =>
-          Left(ToolFormError(s"Failed to parse project configuration [$projectDefinitionFile]: [${e.getMessage}]"))
+          Left(NonEmptyList.of(ToolFormError(s"Failed to parse project configuration [$projectDefinitionFile]: [${e.getMessage}]")))
+      }
+    }
+
+  private def validateProjectResult(project: Project): Either[NonEmptyList[ToolFormError], Project] =
+    project.topology.endpoints.toList
+      .traverseU({
+        case (endpointId, endpoint) =>
+          validateEndpointAgainstComponents(endpointId, endpoint, project.components)
+      })
+      .map { _ =>
+        project
       }
+      .toEither
+
+  private def validateEndpointAgainstComponents(endpointId: String, endpoint: Endpoint, components: Map[String, Component]): Validated[NonEmptyList[ToolFormError], Endpoint] = {
+    val targetComponent = components.get(endpoint.target)
+    targetComponent match {
+      case Some(component) =>
+        val targetPort        = endpoint.portMapping.targetPort
```

Comment:
You could extract this into a function then use it as a guard on the pattern:
```
case Some(component) if targetIsValid(...) =>
case Some(component)                                =>
case _                                                             =>
```

Reply:
I had a feeling that this could be cleaned up 👍